lucene/core/src/java/org/apache/lucene/index/IndexWriter.java

```diff
@@ -941,6 +969,11 @@ public IndexWriter(Directory d, IndexWriterConfig conf) throws IOException {
     // obtain the write.lock. If the user configured a timeout,
     // we wrap with a sleeper and this might take some time.
     writeLock = d.obtainLock(WRITE_LOCK_NAME);
+    if (config.getIndexSort() != null && leafSorter != null) {
+      throw new IllegalArgumentException(
+          "[IndexWriter] can't use index sort and leaf sorter at the same time!");
```

Comment:
@mikemccand Thank you for your review! From the discussion on the Jira ticket, we also wanted to use writer's `leafSorter` during merging for arranging docs in  a merged segment (by putting first docs from a segment with highest sort values according to `leafSorter`).

This will be in conflict with `indexSorter`, as if provided [it will arrange merged docs according to its sort](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/index/MergeState.java#L211).

Reply:
I think it's OK - that MultiSorter is sorting (by index sort) documents in several segments being merged; it will control the order of the documents within the new segment, but shouldn't have any influence on or conflict with the order of the segments being merged

Reply:
@msokolov Thank you for your feedback and explanation. Sorry, I am still not super clear about this point. It seems to me as  `indexSorter` maps each leaf's documents into the merged segment according to its sort,  the same way `leafSorter` will map each leaf's documents into the merged segment according to its sort (given several merging segments, in the merged segment the first docs should be docs from a segment with highest sort values..)

But I am ok to remove this check in this PR, as this PR is not concerned with merging, and follow up on this point in the following PR. 

Addressed in 7ddff6775c8

Reply:
OK, maybe I misunderstood the intent. Perhaps an example would clarify. Say that we have three segments, A, B , C containing documents `A={0, 3, 6}; B={1, 4, 7}; C={2, 5, 8}`, where the documents are understood to have a single field with the value shown, and the index sort is ordered in the natural way.  Without this change, if we merged A and B, we'd get a new segment `A+B={0, 1, 3, 4, 6, 7}`. Now suppose there is no index sort (and the documents just "happen" to be in the index in the order given above, for the sake of the example), and we apply a `leafSorter` that sorts by the minimum value of any document in the segment (I guess it could be any sort of aggregate over the segment?), then we would get `A+B={0, 3, 6, 1, 4, 7}`. Now if we apply both sorts, we would get the same result as in the first case, right? I'm still unclear how the conflict arises.

Reply:
Hmm I do see where you said we want to use `leafSorter` to sort documents. I guess I might challenge that design in favor of building on the `indexSort` we already have? But perhaps this is better sorted out in the context of a later PR, as you suggested.

Reply:
@msokolov Thank you for  the clarification.  Indeed, it is much clear with an example you provided. Looks like we need to think  and discuss more about merging scenario, may be in the next PR or Jira ticket.